File: q19_remove nth node from end of list/solution.py
'''
Given a linked list, remove the n-th node from the end of list and return its head.

Example:

Given linked list: 1->2->3->4->5, and n = 2.

After removing the second node from the end, the linked list becomes 1->2->3->5.
Note:

Given n will always be valid.

Follow up:

Could you do this in one pass?
'''
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        last = head
        for i in range(n):
            last=last.next
        first = head
        if not last:
            return(head.next)
        while(last.next):
            last = last.next
            first = first.next
        first.next = first.next.next
        return(head)

File: q19_remove nth node from end of list/solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode start=new ListNode(0),slow=start,fast=start;
        start.next=head;
        for(int i=0;i<n;i++) fast=fast.next;
        while(fast.next!=null) {fast=fast.next;slow=slow.next;}
        slow.next=slow.next.next;
        return start.next;
}
}
